src/pagination.rs

```diff
@@ -0,0 +1,279 @@
+use cobalt_model::pagination_config::PaginationCfg;
+use cobalt_model::permalink;
+use cobalt_model::SortOrder;
+use document;
+use document::Document;
+use liquid;
+use std::cmp::Ordering;
+use std::ops::Range;
+use std::path::PathBuf;
+
+pub(crate) fn generate_paginators(
+  doc: &mut Document,
+  posts_data: &[liquid::Value],
+  config: &mut PaginationCfg,
+) -> Vec<liquid::Object> {
+  let mut all_posts = posts_data.to_vec();
+  match config.include.as_str() {
+    "all" => {
+      sort_posts(&mut all_posts, &config);
+      create_all_paginators(&mut all_posts, &doc, &config)
+    }
+    _ => {
+      unreachable!();
+    }
+  }
+}
+
+fn extract_page_path_from(p: &liquid::Object) -> String {
+  p.get("page_path")
+    .expect("Should have a page_path")
+    .as_scalar()
+    .expect("Should be a scalar")
+    .to_str()
+    .into_owned()
+}
+
+pub(crate) fn extract_page_path(p: &liquid::Object, config: &PaginationCfg) -> PathBuf {
+  PathBuf::from(match config.include.as_str() {
+    "all" => extract_page_path_from(&p),
+    _ => "".to_owned(),
+  })
+}
+
+fn extract_value(a: &liquid::Value, key: &String) -> Option<liquid::Scalar> {
+  a.as_object().map_or(None, |attr| {
+    attr.get(key).map_or(None, |sort_key| {
+      sort_key
+        .as_scalar()
+        .map_or(None, |value| Some(value.clone()))
+    })
+  })
+}
+
+// sort posts by multiple criteria
+fn sort_posts(posts: &mut Vec<liquid::Value>, config: &PaginationCfg) {
+  posts.sort_by(|a, b| {
+    let keys = config.sort_by.clone();
+    let mut cmp = Ordering::Less;
+    for k in keys {
+      cmp = if let Some(a) = extract_value(a, &k) {
+        if let Some(b) = extract_value(b, &k) {
+          match config.order {
+            SortOrder::Desc => b.partial_cmp(&a).unwrap_or(Ordering::Equal),
+            SortOrder::Asc => a.partial_cmp(&b).unwrap_or(Ordering::Equal),
+            SortOrder::None => unreachable!("We should always have a sort order default value"),
```

Comment:
And do you protect against someone setting it?

Ideally, that would be an error when converting the config to the in-memory model

Reply:
In the reading of the frontmatter configuration, I'm defaulting to "Desc" if something's wrong

Reply:
What if someone asks for `None`?

Reply:
Good question… I don't know ^^'  Defaulting to `Desc`? (As it's the default value I choose for the `order` field)

Reply:
We should probably error if someone explicitly asks for an unsupported value